src/ir/auth_logic/ast_construction.cc

```diff
@@ -0,0 +1,122 @@
+  datalog::Sign sign = context.NEG() == nullptr ? datalog::Sign::kPositive
+                                                : datalog::Sign::kNegated;
+  return datalog::Predicate(context.ID()->getText(), args, sign);
+}
+
+static BaseFact ConstructVerbphrase(
+    Principal left_principal, AuthLogicParser::VerbphraseContext& context) {
+  if (auto* ctx = dynamic_cast<AuthLogicParser::PredphraseContext*>(
+          &context)) {  // PredicateContext
+    return BaseFact(
+        Attribute(left_principal, ConstructPredicate(*ctx->predicate())));
+  }
+  auto* ctx = CHECK_NOTNULL(dynamic_cast<AuthLogicParser::ActsAsPhraseContext*>(
+      &context));  // ActsAsPhraseContext
+  return BaseFact(
+      CanActAs(left_principal, ConstructPrincipal(*ctx->principal())));
+}
+
+static BaseFact ConstructFlatFact(AuthLogicParser::FlatFactContext& context) {
+  if (auto* ctx = dynamic_cast<AuthLogicParser::PrinFactContext*>(&context)) {
+    return ConstructVerbphrase(ConstructPrincipal(*ctx->principal()),
+                               *ctx->verbphrase());
+  }
+  auto* ctx = CHECK_NOTNULL(dynamic_cast<AuthLogicParser::PredFactContext*>(
+      &context));  // predFactContext
+  return BaseFact(ConstructPredicate(*ctx->predicate()));
+}
+
+static Fact ConstructFact(AuthLogicParser::FactContext& context) {
+  if (auto* flat_fact_fact_ctx =
+          dynamic_cast<AuthLogicParser::FlatFactFactContext*>(
+              &context)) {  // FlatFactFactContext
+    return Fact(ConstructFlatFact(*flat_fact_fact_ctx->flatFact()));
+  }
+  auto* ctx = CHECK_NOTNULL(
+      dynamic_cast<AuthLogicParser::CanSayFactContext*>(&context));
+  return ConstructFact(*ctx->fact());
+}
+
+static Query ConstructQuery(AuthLogicParser::QueryContext& context) {
+  return Query(context.ID()->getText(),
+               ConstructPrincipal(*context.principal()),
+               ConstructFact(*context.fact()));
+}
+
+static Program ConstructProgram(AuthLogicParser::ProgramContext& context) {
+  std::vector<SaysAssertion> says_assertions;
+  std::vector<Query> queries = utils::MapIter<Query>(
+      context.query(), [](AuthLogicParser::QueryContext* statement) {
+        return ConstructQuery(*statement);
+      });
+  return Program(std::move(says_assertions), std::move(queries));
+}
+
+/// This function produces an abstract syntax tree (AST) rooted with a program
+/// node when given the textual representation of a program.
+Program ParseProgram(absl::string_view prog_text) {
+  // Provide the input text in a stream
+  antlr4::ANTLRInputStream input(prog_text);
+  // Creates a lexer from input
+  AuthLogicLexer lexer(&input);
+  // Creates a token stream from lexer
+  antlr4::CommonTokenStream tokens(&lexer);
+  // Creates a parser from the token stream
+  AuthLogicParser parser(&tokens);
+  // program_context points to the root of the parse tree
+  AuthLogicParser::ProgramContext* program_context = parser.program();
+  return ConstructProgram(*program_context);
```

Comment:
Need a `CHECK_NOTNULL` in here?

Reply:
The following might be a good pattern to follow:
```C++
  AuthLogicParser::ProgramContext& program_context = *CHECK_NOTNULL(parser.program());
  return ConstructProgram(program_context);
```